File: framework/test2/validate/list/list.py
#!/usr/bin/env python3
from collections import Counter, defaultdict, deque
import heapq
import json
import logging
import math
import random
import time
from typing import *
import sys

from soda.leetcode.bitree import *
from soda.leetcode.graph import *
from soda.leetcode.linklist import *
from soda.leetcode.nest import *
from soda.unittest.util import init_logging

logger = logging.getLogger(__name__)

# step [1]: implement class Solution
# class Solution: pass
class Solution:
    def permutation(self, chars: List[str], n: int) -> List[str]:
        res = []
        buf = [None] * n
        self.solve(chars, 0, buf, 0, res)
        return res

    def solve(self, chars, i, buf, j, res):
        if j == len(buf):
            res.append(''.join(buf))
            return
        for k in range(i, len(chars)):
            temp = chars[i]
            chars[i] = chars[k]
            chars[k] = temp
            buf[j] = chars[i]
            self.solve(chars, i+1, buf, j+1, res)
            temp = chars[i]
            chars[i] = chars[k]
            chars[k] = temp

if __name__ == '__main__':
    init_logging()
    from soda.unittest import Validators
    from soda.unittest.work import TestWork

    # step [2]: setup function
    # Attention! FUNCTION must use type hint, including arguments and return type
    work = TestWork(Solution().permutation)
    # OR use struct tester
    # work = TestWork.forStruct(CLASS)

    # step [3]: setup other options
    work.validator = Validators.forList(False)
    work.compareSerial = True
    work.run()

File: framework/test2/validate/list/list.rb
# step [1]: implement solution function
# @param {String[]} chars
# @param {Integer} n
# @return {String[]}
def permutation(chars, n)
  res = []
  buf = Array.new(n, '')
  solve(chars, 0, buf, 0, res)
  res
end

def solve(chars, i, buf, j, res)
  if j == buf.size
    res << buf.join
    return
  end
  (i...chars.size).each { |k|
    chars[i], chars[k] = chars[k], chars[i]
    buf[j] = chars[i]
    solve(chars, i+1, buf, j+1, res)
    chars[i], chars[k] = chars[k], chars[i]
  }
end

if __FILE__ == $0
  require 'soda/unittest/work'
  include Soda::Unittest
  # step [2]: setup function/return/arguments
  work = TestWork.create(method(:permutation))
  # work = ns::TestWork.forStruct(CLASS)
  work.validator = ->(e, r) { Validators.for_array('String', false) }
  work.compare_serial = true
  puts work.run(ARGF.read)
end

